src/uct/base/uct_mem.c

```diff
@@ -334,14 +334,54 @@ ucs_status_t uct_mem_free(const uct_allocated_memory_t *mem)
     }
 }
 
+static int uct_iface_is_allowed_alloc_method(uct_base_iface_t *iface,
+                                             uct_alloc_method_t method)
+{
+    unsigned i;
+
+    for (i = 0; i < iface->config.num_alloc_methods; i++) {
+        if (iface->config.alloc_methods[i] == method) {
+            return 1;
+        }
+    }
+
+    return 0;
+}
+
 ucs_status_t uct_iface_mem_alloc(uct_iface_h tl_iface, size_t length, unsigned flags,
                                  const char *name, uct_allocated_memory_t *mem)
 {
-    uct_base_iface_t *iface = ucs_derived_of(tl_iface, uct_base_iface_t);
-    void *address           = NULL;
+    static uct_alloc_method_t method_md = UCT_ALLOC_METHOD_MD;
+    uct_base_iface_t *iface             = ucs_derived_of(tl_iface,
+                                                         uct_base_iface_t);
+    void *address                       = NULL;
     uct_md_attr_t md_attr;
     ucs_status_t status;
     uct_mem_alloc_params_t params;
+    unsigned num_alloc_methods;
+    uct_alloc_method_t *alloc_methods;
+
+    status = uct_md_query(iface->md, &md_attr);
+    if (status != UCS_OK) {
+        goto err;
+    }
+
+    if (!(md_attr.cap.flags & UCT_MD_FLAG_REG) &&
+        uct_iface_is_allowed_alloc_method(iface, UCT_ALLOC_METHOD_MD)) {
+        /* If MD does not support registration, allow only the MD method */
+        alloc_methods     = &method_md;
+        num_alloc_methods = 1;
+    } else if (!(md_attr.cap.flags & UCT_MD_FLAG_REG)) {
+        /* If MD does not support registration and MD method is not in list
+         * return error */
+        ucs_assert(!uct_iface_is_allowed_alloc_method(iface,
+                                                      UCT_ALLOC_METHOD_MD));
```

Comment:
imo this is excessive as it is guaranteed by the if conditions

Reply:
ok, removed